Author: Christian Heimes <christian@python.org>
Description: bpo-43794: OpenSSL 3.0.0: set OP_IGNORE_UNEXPECTED_EOF by default (GH-25309)
Origin: upstream, https://github.com/python/cpython/commit/3309113d6131e4bbac570c4f54175ecca02d025a

Index: python2.7-2.7.18/Doc/library/ssl.rst
===================================================================
--- python2.7-2.7.18.orig/Doc/library/ssl.rst
+++ python2.7-2.7.18/Doc/library/ssl.rst
@@ -760,6 +760,12 @@ Constants
 
    .. versionadded:: 2.7.9
 
+.. data:: OP_IGNORE_UNEXPECTED_EOF
+
+   Ignore unexpected shutdown of TLS connections.
+
+   This option is only available with OpenSSL 3.0.0 and later.
+
 .. data:: HAS_ALPN
 
    Whether the OpenSSL library has built-in support for the *Application-Layer
Index: python2.7-2.7.18/Lib/test/test_ssl.py
===================================================================
--- python2.7-2.7.18.orig/Lib/test/test_ssl.py
+++ python2.7-2.7.18/Lib/test/test_ssl.py
@@ -84,6 +84,7 @@ OP_SINGLE_DH_USE = getattr(ssl, "OP_SING
 OP_SINGLE_ECDH_USE = getattr(ssl, "OP_SINGLE_ECDH_USE", 0)
 OP_CIPHER_SERVER_PREFERENCE = getattr(ssl, "OP_CIPHER_SERVER_PREFERENCE", 0)
 OP_ENABLE_MIDDLEBOX_COMPAT = getattr(ssl, "OP_ENABLE_MIDDLEBOX_COMPAT", 0)
+OP_IGNORE_UNEXPECTED_EOF = getattr(ssl, "OP_IGNORE_UNEXPECTED_EOF", 0)
 
 
 def handle_error(prefix):
@@ -839,7 +840,8 @@ class ContextTests(unittest.TestCase):
         # SSLContext also enables these by default
         default |= (OP_NO_COMPRESSION | OP_CIPHER_SERVER_PREFERENCE |
                     OP_SINGLE_DH_USE | OP_SINGLE_ECDH_USE |
-                    OP_ENABLE_MIDDLEBOX_COMPAT)
+                    OP_ENABLE_MIDDLEBOX_COMPAT |
+                    OP_IGNORE_UNEXPECTED_EOF)
         self.assertEqual(default, ctx.options)
         ctx.options |= ssl.OP_NO_TLSv1
         self.assertEqual(default | ssl.OP_NO_TLSv1, ctx.options)
Index: python2.7-2.7.18/Modules/_ssl.c
===================================================================
--- python2.7-2.7.18.orig/Modules/_ssl.c
+++ python2.7-2.7.18/Modules/_ssl.c
@@ -2313,6 +2313,10 @@ context_new(PyTypeObject *type, PyObject
 #ifdef SSL_OP_SINGLE_ECDH_USE
     options |= SSL_OP_SINGLE_ECDH_USE;
 #endif
+#ifdef SSL_OP_IGNORE_UNEXPECTED_EOF
+    /* Make OpenSSL 3.0.0 behave like 1.1.1 */
+    options |= SSL_OP_IGNORE_UNEXPECTED_EOF;
+#endif
     SSL_CTX_set_options(self->ctx, options);
 
     /* A bare minimum cipher list without completly broken cipher suites.
@@ -4468,6 +4472,10 @@ init_ssl(void)
     PyModule_AddIntConstant(m, "OP_ENABLE_MIDDLEBOX_COMPAT",
                             SSL_OP_ENABLE_MIDDLEBOX_COMPAT);
 #endif
+#ifdef SSL_OP_IGNORE_UNEXPECTED_EOF
+    PyModule_AddIntConstant(m, "OP_IGNORE_UNEXPECTED_EOF",
+                            SSL_OP_IGNORE_UNEXPECTED_EOF);
+#endif
 
 #if HAVE_SNI
     r = Py_True;
